File: apps/demo-be-rsax/tests/integration/steps/token_management_steps.rs
use cucumber::{given, then, when};

use crate::world::{get_req, json_req, AppWorld};

#[when("alice decodes her access token payload")]
async fn decode_alice_token(world: &mut AppWorld) {
    let bearer = world.bearer();
    let req = get_req("/api/v1/tokens/claims", Some(&bearer));
    world.send(req).await.unwrap();
}

#[then(expr = "the token should contain a non-null {string} claim")]
async fn token_claim_non_null(world: &mut AppWorld, claim: String) {
    let val = world.last_body.get(&claim);
    assert!(
        val.is_some() && !val.unwrap().is_null(),
        "Claim '{claim}' should be non-null in body: {}",
        world.last_body
    );
}

#[when("the client sends GET /.well-known/jwks.json")]
async fn get_jwks(world: &mut AppWorld) {
    let req = get_req("/.well-known/jwks.json", None);
    world.send(req).await.unwrap();
}

#[then(expr = "the response body should contain at least one key in the {string} array")]
async fn jwks_has_keys(world: &mut AppWorld, field: String) {
    let keys = world
        .last_body
        .get(&field)
        .and_then(|v| v.as_array())
        .map(|a| a.len())
        .unwrap_or(0);
    assert!(
        keys >= 1,
        "Expected at least one key in '{field}', got {keys}"
    );
}

#[then("alice's access token should be recorded as revoked")]
async fn token_is_revoked(world: &mut AppWorld) {
    // Try to use the token to access a protected endpoint
    let token = world.auth_token.clone().unwrap_or_default();
    let req = get_req("/api/v1/users/me", Some(&format!("Bearer {token}")));
    world.send(req).await.unwrap();
    assert_eq!(
        world.last_status, 401,
        "Expected 401 for revoked token, got {}",
        world.last_status
    );
}

#[given("alice has logged out and her access token is blacklisted")]
async fn alice_logged_out_blacklisted(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    let body = format!(r#"{{"access_token": "{token}"}}"#);
    let req = json_req("POST", "/api/v1/auth/logout", &body, None);
    world.send(req).await.unwrap();
}

#[given(
    regex = r#"the admin has disabled alice's account via POST /api/v1/admin/users/\{alice_id\}/disable"#
)]
async fn admin_disabled_alice(world: &mut AppWorld) {
    let admin_bearer = world.admin_bearer();
    let alice_id = world.alice_id.map(|id| id.to_string()).unwrap_or_default();
    if !alice_id.is_empty() && !admin_bearer.is_empty() {
        let req = json_req(
            "POST",
            &format!("/api/v1/admin/users/{alice_id}/disable"),
            r#"{"reason": "test"}"#,
            Some(&admin_bearer),
        );
        world.send(req).await.unwrap();
    }
}

File: apps/demo-be-python-fastapi/tests/integration/steps/token_management_steps.py
"""BDD step definitions for token management feature."""

import base64
import json

from pytest_bdd import given, parsers, scenarios, then, when

from tests.integration.conftest import GHERKIN_ROOT
from tests.integration.service_client import FakeResponse, ServiceClient
from tests.integration.steps.security_steps import _ADMIN_PASSWORD, _register_and_promote_admin

scenarios(str(GHERKIN_ROOT / "token-management" / "tokens.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    '"alice" has logged in and stored the access token',
    target_fixture="alice_tokens",
)
def alice_login_for_tokens(client: ServiceClient, registered_user: dict) -> dict:
    return client.login_user("alice", _PASSWORD)


@given("alice has logged out and her access token is blacklisted")
def alice_logout_blacklisted(client: ServiceClient, alice_tokens: dict) -> None:
    client.post_logout(f"Bearer {alice_tokens['access_token']}")


@given(
    parsers.parse('an admin user "{username}" is registered and logged in'),
    target_fixture="admin_tokens",
)
def admin_login_for_tokens(client: ServiceClient, username: str) -> dict:
    user_data = _register_and_promote_admin(client, username, _ADMIN_PASSWORD)
    tokens = client.login_user(username, _ADMIN_PASSWORD)
    return {**tokens, "id": user_data["id"]}


@given("the admin has disabled alice's account via POST /api/v1/admin/users/{alice_id}/disable")
def admin_disable_alice(
    client: ServiceClient, registered_user: dict, admin_tokens: dict
) -> None:
    resp = client.post_admin_disable_user(
        registered_user["id"],
        f"Bearer {admin_tokens['access_token']}",
        reason="Test",
    )
    assert resp.status_code == 200, f"Disable failed: {resp.text}"


# --- When steps ---


@when("alice decodes her access token payload", target_fixture="decoded_payload")
def alice_decode_token(alice_tokens: dict) -> dict:
    token = alice_tokens["access_token"]
    parts = token.split(".")
    assert len(parts) == 3, "Invalid JWT structure"
    payload_b64 = parts[1]
    padding = 4 - len(payload_b64) % 4
    if padding != 4:
        payload_b64 += "=" * padding
    return json.loads(base64.urlsafe_b64decode(payload_b64).decode("utf-8"))


@when("the client sends GET /.well-known/jwks.json", target_fixture="response")
def get_jwks(client: ServiceClient) -> FakeResponse:
    return client.get_jwks()


@when("alice sends POST /api/v1/auth/logout with her access token", target_fixture="response")
def alice_logout_tokens(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    return client.post_logout(f"Bearer {alice_tokens['access_token']}")


@when(
    "the client sends GET /api/v1/users/me with alice's access token",
    target_fixture="response",
)
def get_me_with_alice_token(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    return client.get_me(f"Bearer {alice_tokens['access_token']}")


# --- Then steps ---


@then(parsers.parse('the token should contain a non-null "{claim}" claim'))
def check_token_claim(decoded_payload: dict, claim: str) -> None:
    assert claim in decoded_payload, f"Claim '{claim}' not in token: {decoded_payload}"
    assert decoded_payload[claim] is not None, f"Claim '{claim}' is null"


@then('the response body should contain at least one key in the "keys" array')
def check_jwks_keys(response: FakeResponse) -> None:
    body = response.json()
    assert "keys" in body, f"No 'keys' field in JWKS response: {body}"
    assert len(body["keys"]) > 0, "JWKS keys array is empty"


@then("alice's access token should be recorded as revoked")
def check_alice_token_revoked(client: ServiceClient, alice_tokens: dict) -> None:
    resp = client.get_me(f"Bearer {alice_tokens['access_token']}")
    assert resp.status_code == 401, f"Expected 401 for revoked token, got {resp.status_code}"
